clients/js/src/compat.ts

```diff
@@ -0,0 +1,289 @@
+  { method, params }: Web3ReqArgs,
+  signer: JsonRpcSigner,
+  opts: ProviderOptions | undefined,
+): Promise<[unknown[], MetaEncryptedX25519DeoxysII?]> {
+  if (!Array.isArray(params)) return [[params]];
+
+  const enc = async (data: BytesLike, chainId?: number) =>
+    encrypt(data, chainId ?? (await signer.getChainId()), opts);
+
+  if (method.endsWith('sendRawTransaction')) {
+    return [[enc(params[0]), ...params.slice(1)]];
+  }
+
+  if (
+    (method.endsWith('call') || method.endsWith('estimateGas')) &&
+    (await callNeedsSigning(params[0]))
+  ) {
+    const chainId = await signer.getChainId();
+    const [encData, meta] = await enc(params[0].data ?? [], chainId);
+    params[0].data = encData;
+    return [
+      [prepareSignedCall(params[0], signer, { chainId }), ...params.slice(1)],
+      meta,
+    ];
+  }
+
+  if (
+    method.endsWith('sendTransaction') ||
+    method.endsWith('signTransaction')
+  ) {
+    if (isBytesLike(params[0].data)) {
+      params[0].data = (await enc(params[0]))[0];
+    }
+    return [params];
+  }
+
+  return [params];
+}
+
+async function encrypt(
+  data: BytesLike,
+  chainId: number,
+  opts: ProviderOptions | undefined,
+): Promise<[string, MetaEncryptedX25519DeoxysII?]> {
+  if (data.length === 0) return [hexlify(data)];
+
+  const [packed, meta] = await callformat.encodeCall(
+    {
+      method: '',
+      body: arrayify(data),
+    },
+    CALLFORMAT_ENCRYPTED_X25519DEOXYSII,
+    { publicKey: await fetchRTPubKey(chainId, opts) },
+  );
+  return [hexlify(cbor.encode(packed)), meta as MetaEncryptedX25519DeoxysII];
+}
+
+async function decrypt(
+  data: Uint8Array,
+  meta: MetaEncryptedX25519DeoxysII,
+): Promise<string> {
+  const callResult = cbor.decode(data);
+  const { ok, fail, unknown, ...trulyUnknown } = await callformat.decodeResult(
+    callResult,
+    CALLFORMAT_ENCRYPTED_X25519DEOXYSII,
+    meta,
+  );
+
+  if (ok) {
+    if (typeof ok === 'string') return ok;
+    const err = new Error(`unexpected EVM response: ${ok}`);
+    (err as any).response = ok;
+    throw err;
+  }
+
+  if (fail) {
+    const msg = fail.message
+      ? fail.message
+      : `call failed in ${fail.module} with code ${fail.code}`;
+    const err = new Error(msg);
+    (err as any).cause = fail;
+    throw err;
+  }
+
+  if (unknown) {
+    const err = new Error(`unexpected EVM response: ${unknown}`);
+    (err as any).response = unknown;
+    throw err;
+  }
+
+  const err = new Error('unknown response from runtime');
+  (err as any).response = trulyUnknown;
+  throw err;
+}
+
+async function fetchRTPubKey(
+  chainId: number,
+  opts?: ProviderOptions,
+): Promise<KeyManagerSignedPublicKey> {
+  const cached = PK_CACHE.get(chainId);
+  if (cached) return cached;
+
+  const rtId = opts?.runtimeId ?? inferRuntimeId(chainId);
+  const rpc = opts?.rpc ?? inferRpcUrl(chainId);
+  const { public_key: publicKey } = await new CoreWrapper(arrayify(rtId))
+    .queryCallDataPublicKey()
```

Comment:
Should this somehow be exposed via a Web3 REST API extension to avoid the need for two separate connections?

Reply:
it absolutely should, but we need to cut an alpha of this lib asap